### ISSUE TITLE
Errors during get/set (Cannot read property [......] of undefined)

### ISSUE THREAD

**Issue by giovannipiller** [ractivejs/ractive#2689]
**Description of the bug:**

Attached, you'll find a couple errors & stacktraces that I've been experiencing on my application.

**Background of the app:** it's a single-page-application, with a decent number of components created and destroyed during its life. These errors seem reproducible whenever some specific components are tore down.

As you can see from the stacktraces, the source of the errors is either a `get` or `set`.

Is it possible that these get/set are somehow happening _after_ the component has been tore down? That was my first thought.

Building a fiddle out of my code seems quite a challenge, but I can invest some time if the stacktraces are of no use. Should you have any tip or specific debugs you'd like to see, just ask.

**Versions affected:**
0.8 and 0.8.1-build-2 (https://github.com/ractivejs/ractive.git#5aa6d17d006398e0d63feeddda6543548f0db63c)
No issue on 0.7.3.

**Reproduction:**

**Code:**

Stacktrace:
```
ractive.js?ee9c:8423 Uncaught TypeError: Cannot read property 'get' of undefined(…)
get	@	ractive.js?ee9c:8423
Ractive$get	@	ractive.js?ee9c:2915
MyFunction	@	MyCode?1ab5:215
```

Relevant code:
```
ExpressionProxy.prototype.get = function get ( shouldCapture ) {
	return this.computation.get( shouldCapture );
};
```

`this.computation` is `undefined`

----

Stacktrace:
```
ractive.js?ee9c:9755 Uncaught TypeError: Cannot read property 'wrappers' of undefined(…)
teardown	@	ractive.js?ee9c:9755
adapt	@	ractive.js?ee9c:2284
mark$1	@	ractive.js?ee9c:2585
mark	@	ractive.js?ee9c:1973
applyValue	@	ractive.js?ee9c:2382
set	@	ractive.js?ee9c:2681
set$2	@	ractive.js?ee9c:15955
Ractive$set	@	ractive.js?ee9c:15937
MyFunction	@	MyCode?5914:368
```

Relevant code:
```
 ArrayWrapper.prototype.teardown = function teardown () {
  	var array, storage, wrappers, instances, index;

  	array = this.value;
  	storage = array._ractive;
  	wrappers = storage.wrappers;
```

`storage` is undefined

**Comment by giovannipiller** [comment]
I'm investigating a bit more, as the issue seems to be somehow connected to decorators. Trying to build a fiddle. Will update soon.

**Comment by giovannipiller** [comment]
Isolating the issue is proving to be quite difficult.
What I'm pretty sure now, is that there are cases in which Decorator's teardown functions are not executed when the associated element is removed.

Pretty strange, as I've been experiencing this with pretty simple interactions (show/hide, plain decorators with just console.logs).

**This seems to be somehow caused by `{{#with}}..{{/with}}`.**

In my app, decorators instantiated like this, teardown properly whenever the `<div>` is removed:
```
<div decorator="myDecorator"></div>
{{#with {anyDataReally: true} }}{{/with}}
```

In my app, decorators instantiated like this, will NOT fire the teardown:
```
{{#with {anyDataReally: true} }}
   <div decorator="myDecorator"></div>
{{/with}}
```

Using the new syntax `as-myDecorator` doesn't give different results.

All my attempts to replicate the issue on jsfiddle have failed so far.

**Comment by evs-chris** [comment]
Is your app available publicly somewhere that I could run it with a set of steps that cause the error? If so, I could probably tease out a test case. It sounds like you're porting a decent sized app, and I'd like to get all of the issues you run across patched up for 0.8.1.

**Comment by giovannipiller** [comment]
Unfortunately it's not publicly available.

I'm trying to narrow it down as much as possible. Right now I can reproduce it **only if the template of the component is something like this**:
```
<div>
    {{#with { example: true } }}
      <div decorator="tooltipDecorator"></div>
    {{/with}}
</div>
```

With the decorator working properly if no `<div>` wraps the `{{#with}}`
```
{{#with { example: true } }}
    <div decorator="tooltipDecorator"></div>
{{/with}}
```

Hopefully I'll be able to further pinpoint the issue.
Excuse the amount of comments, but I'm trying to keep track of everything while I'm performing other tasks.

**Comment by giovannipiller** [comment]
I'm starting to think that it might be caused by something during my build process.

I use Babel + Webpack, and stripping the app to the bare minimum (2 components and a few lines of templates) didn't help in isolating the issue.

Tried the same code with a fresh ES5  build of Ractive and the issue isn't there.

Don't think it's Webpack, since I removed all the optional loaders (ractive-loader too, I'm passing templates as strings). Will try to upgrade Babel and see how it goes from there.

**Comment by giovannipiller** [comment]
I ended up building a sample project where I was able to reproduce the issue.
Should be to the bare minimum.

Hope this helps.

https://github.com/giovannipiller/ractive-webpack-babel-issue-2689

**Comment by evs-chris** [comment]
I think this is an issue with trying to bundle ractive from the src repo, which doesn't have an appropriate package.json. If I change the package.json in that repo to reference `"ractive": "0.8.1-build-2"` and the require in the app.js to `require("ractive")`, everything builds and works as expected when run.

If you want to pull a specific version from git, I think you would have to use a commit on the build branch.

**Comment by giovannipiller** [comment]
Thanks a lot! I'll give a try next week!

**Comment by giovannipiller** [comment]
Ok, it looks like the "build-related" issues are solved by using an actual release.

> It sounds like you're porting a decent sized app, and I'd like to get all of the issues you run across patched up for 0.8.1.

To give you an idea, I have now spotted a few more issues so far:
- some of my decorators teardown methods trigger twice (might be an issue on my side, not sure yet)
- there's a "Maximum call stack size exceeded" due to a recursive loop in the shuffle, triggered by a `this.merge(...)` somewhere in the code. Using `this.set(...)` everything is fine.

Will open new tickets as soon as I can give you more details, but since there's quite a number of moving parts involved, it might take a few days.

Again, thanks for the help.
